### ISSUE TITLE
Update to Fody Version 2

### ISSUE THREAD

**Issue by SimonCropp** [kswoll/ReactiveUI.Fody#40]
Fody Version 2 is out and you will need to do an update for your addin to be usable with it

Specifically:

 * Update to the newest Fody.Cecil nuget
 * Change the min dependency range for Fody to be at least 2.0

 Please ping me if u need help

**Comment by kswoll** [comment]
Thanks for the new version, @SimonCropp!  I actually have a -Pre version already going -- doing some testing to make sure everything is kosher.  Also, in one of my other plugins, I'm noticing two problems with the new version:

1.  Getting this error when creating a whole new method:
```
Value does not fall within the expected range.
StackTrace:
   at Mono.Cecil.Pdb.ISymUnmanagedWriter2.DefineLocalVariable2(String name, Int32 attributes, Int32 sigToken, Int32 addrKind, Int32 addr1, Int32 addr2, Int32 addr3, Int32 startOffset, Int32 endOffset)
   at Mono.Cecil.Pdb.NativePdbWriter.DefineScope(ScopeDebugInformation scope, MethodDebugInformation info, MetadataToken& import_parent)
   at Mono.Cecil.Pdb.NativePdbWriter.Write(MethodDebugInformation info)
   at Mono.Cecil.Cil.CodeWriter.WriteResolvedMethodBody(MethodDefinition method)
   at Mono.Cecil.Cil.CodeWriter.WriteMethodBody(MethodDefinition method)
   at Mono.Cecil.MetadataBuilder.AddMethod(MethodDefinition method)
   at Mono.Cecil.MetadataBuilder.AddMethods(TypeDefinition type)
   at Mono.Cecil.MetadataBuilder.AddType(TypeDefinition type)
   at Mono.Cecil.MetadataBuilder.AddTypes()
   at Mono.Cecil.MetadataBuilder.BuildTypes()
   at Mono.Cecil.MetadataBuilder.BuildModule()
   at Mono.Cecil.MetadataBuilder.BuildMetadata()
   at Mono.Cecil.ModuleWriter.<>c.<BuildMetadata>b__1_0(MetadataBuilder builder, MetadataReader _)
   at Mono.Cecil.ModuleDefinition.Read[TItem,TRet](TItem item, Func`3 read)
   at Mono.Cecil.ModuleWriter.BuildMetadata(ModuleDefinition module, MetadataBuilder metadata)
   at Mono.Cecil.ModuleWriter.WriteModuleTo(ModuleDefinition module, Disposable`1 stream, WriterParameters parameters)
   at Mono.Cecil.ModuleDefinition.Write(String fileName, WriterParameters parameters)
   at InnerWeaver.WriteModule() in C:\projects\fody\FodyIsolated\ModuleWriter.cs:line 18
   at InnerWeaver.Execute() in C:\projects\fody\FodyIsolated\InnerWeaver.cs:line 90
```

2. Fody seems to keep a file lock on pdbs so after the first compile, I have to exit Visual Studio or manually kill MsBuild.exe each time:

> Cannot open 'C:\dev\plangrid-windows\lib\obj\Debug\PlanGrid.Library.pdb' for writing -- 'The process cannot access the file 'C:\dev\plangrid-windows\lib\obj\Debug\PlanGrid.Library.pdb' because it is being used by another process.'

So if you have any insights on what to do about these two issues, any help would be appreciated.  I _suspect_ problem 1) is likely an issue that will need to be addressed by Mono.Cecil.  Issue 2) I'm less sure about -- it's very consistent for me, but I'm not sure if there's a bug in Fody or Mono.Cecil that is keeping a file lock on those pdbs, but definitely only happens with v2.

**Comment by SimonCropp** [comment]
What versions of fody and weavers are u using. And can u repro in a stand alone solution?

**Comment by SimonCropp** [comment]
I would also appreciate a review of the pdb reading and writing code if u have time

**Comment by kswoll** [comment]
@SimonCropp I've attached a small zip that is a minimal example that reproduces both issues.  I think the file locking issue is related to the fody build failing -- probably something not getting disposed in that scenario?

I'm using Fody 2.0.2.  I'm using the weaver:  SexyProxy.1.0.65-beta1 (this is my own package).  The precise nature of the fody build failure is related to Mono.Cecil trying to round trip debug information for local variables in a generated method.

I'm happy to review any pdb reading/writing code -- can you elaborate a bit more as to what code you are referring to?

[FodyCecilProxyIssue.zip](https://github.com/kswoll/ReactiveUI.Fody/files/952276/FodyCecilProxyIssue.zip)

**Comment by SimonCropp** [comment]
@kswoll can u repo the issue in your unit tests on SexyProxy?

**Comment by SimonCropp** [comment]
btw when u push your beta to nuget dont forget to update the min nuspec version https://github.com/kswoll/ReactiveUI.Fody/blob/master/Nuget/ReactiveUIFody.nuspec#L16

**Comment by kswoll** [comment]
Thanks for the heads up about the nuspect, @SimonCropp.  Also, I managed to fix that stack trace I mentioned above by adding this line:

    Method.DebugInformation.Scope = null;

You see, on the next line I do:

    Method.Body = body = new MethodBody(Method);

At that point, the cached debug information (which contained now obsolete local variables) was inapplicable to the new method body, and hence we got the error.  This way of overwriting the method body didn't cause problems in the past, but it does now presumably due to round-tripping changes.  Here's the fix in context:

https://github.com/kswoll/sexy-proxy/commit/5c1c8837ea9ba6b8a12f4895efe97f5c6d3f68b5#diff-ef9eb9812f2d1b8eb220137c22ea404dR86

Also, since the file locking only happened when this error was happening (on write of the assembly), that issue has since gone away for me.
